node/src/node/app.cpp

```diff
@@ -573,25 +575,42 @@ class cocaine::service::node::app_state_t {
         virtual
         void
         on_abort(const std::error_code& ec, const std::string& reason) {
-            COCAINE_LOG_ERROR(parent.log, "unable to spool app, [{}] {} - {}", ec.value(), ec.message(), reason);
+            const auto p = parent.lock();
+            if (!p) {
+                // TODO: should we inform RT (or logger) that there was nothing to abort?
+                // throw?
+                return;
+            }
+            COCAINE_LOG_ERROR(p->log, "unable to spool app, [{}] {} - {}", ec.value(), ec.message(), reason);
             // Dispatch the completion handler to be sure it will be called in a I/O thread to
             // avoid possible deadlocks.
-            parent.loop->dispatch(std::bind(&app_state_t::cancel, &parent, ec));
+            p->loop->dispatch(std::bind(&app_state_t::cancel, p, ec));
 
-            parent.callback(make_exceptional_future<void>(std::system_error(ec, reason)));
+            p->callback(make_exceptional_future<void>(std::system_error(ec, reason)));
         }
+
         virtual
         void
         on_ready() {
-            COCAINE_LOG_DEBUG(parent.log, "application has been spooled");
-            parent.loop->dispatch(std::bind(&app_state_t::publish, &parent));
+            const auto p = parent.lock();
+            if (!p) {
+                // TODO: should we inform RT (or logger) that application was never cast alive?
+                // Throw?
+                return;
+            }
+
+            COCAINE_LOG_DEBUG(p->log, "application has been spooled");
+            p->loop->dispatch(std::bind(&app_state_t::publish, p));
         }
 
-        spool_handle_t(app_state_t& _parent) :
-            parent(_parent)
+        spool_handle_t(std::shared_ptr<app_state_t> _parent) :
```

Comment:
minor, but looks excess copy here, because there is no weak_ptr(shared_ptr&&) ctor. either take weak_ptr by value or shared_ptr by ref